File: allfiles/App9/tdma.f90
!************************ T D M A *******************************
!                                                               *
!     Solution of a linear system of algebraic equations with   *
!     a tridiagonal matrix of coefficients.(No pivoting)        *
!     Equation no. i :                                          *
!         a(i)*x(i-1) + b(i)*x(i) + c(i)*x(i+1) = d(i),         *
!                                            i = 1,2,...n       *
!                       === USE ===                             *
!                                                               *
!                  call tdma(n,a,b,c,d,x)                       *
!                            or                                 *
!                  call tdma(n,a,b,c,d,d)                       *
!                                                               *
!      In the last case, vector d contains the solution.        *
!                                                               *
!                     === INPUT ===                             *
!                                                               *
!     n ....... integer     . Number of equations               *
!     a(1:n) .. real vector . Lower diagonal.Element A(1)       *
!                                            is not used.       *
!     b(1:n) .. real vector . Main diagonal                     *
!     c(1:n) .. real vector . Upper diagonal.Element C(N)       *
!                                            is not used.       *
!     d(1:n) .. real vector . Right hand side of the system.    *
!                                                               *
!                     === OUTPUT ===                            *
!                                                               *
!     x(1:n) .. real vector . The solution vector               *
!                                                               *
!********************** fortran 90 ******************************

      subroutine tdma(n,a,b,c,d,x)
	  implicit none
      integer, intent(in) :: n
      real, intent(in) :: a(n), c(n)
      real, intent(inout), dimension(n) :: b, d
	  real, intent(out) :: x(n)
	  !  --- Local variables ---
	  integer :: i
	  real :: q
      !  --- Elimination ---
      do i = 2,n
         q = a(i)/b(i - 1)
         b(i) = b(i) - c(i - 1)*q
         d(i) = d(i) - d(i - 1)*q
      end do
      ! --- Backsubstitution ---
      q = d(n)/b(n)
      x(n) = q
      do i = n - 1,1,-1
         q = (d(i) - c(i)*q)/b(i)
         x(i) = q
      end do
      return
      end


File: allfiles/App1/tdma.m
function x = tdma(a,b,c,d)
% Solution of a linear system of algebraic 
% equations with a tri-diagonal matrix of coefficients
% using the Thomas-algorithm (No pivoting).
% Equation no. i :
%  a(i)*x(i-1) + b(i)*x(i) + c(i)*x(i+1) = d(i), i = 1,..,n
%  where n is the number of equations.
%
%         === Input ===
% Vectors may be in either row - or column-form
%
%  a(1:n) .. Lower diagonal. a(1) is not used
%  b(1:n) .. Main diagonal. 
%  c(1:n) .. Upper diagonal. c(n) is not used
%  d(1:n) .. Right hand side of the system.
%
%         === Output ===
%  x(1:n) .. The solution vector

n = length(b);
x = zeros(size(b));
% === Elimination ===
for k = 2:n
   q = a(k)/b(k-1);
   b(k) = b(k) - c(k-1)*q;
   d(k) = d(k) - d(k-1)*q;
end
% === Backsubstitution ===
q = d(n)/b(n);
x(n) = q;
for k = n-1 :-1 :1
   q = (d(k) - c(k)*q)/b(k);
   x(k) = q;
end
